### PR TITLE
[SPARK-24360][SQL] Support Hive 3.0 metastore

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

Hive 3.0 is released. This PR aims to support Hive 3.0 metastore.

## How was this patch tested?

Pass the Jenkins with the updated test cases including 3.0.

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/package.scala [hunk 1]
```diff
@@ -79,7 +79,12 @@ package object client {
       exclusions = Seq("org.apache.curator:*",
         "org.pentaho:pentaho-aggdesigner-algorithm"))
 
-    val allSupportedHiveVersions = Set(v12, v13, v14, v1_0, v1_1, v1_2, v2_0, v2_1, v2_2, v2_3)
+    case object v3_0 extends HiveVersion("3.0.0",
+      exclusions = Seq("org.apache.curator:*",
+        "org.apache.hadoop:hadoop-aws",
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/package.scala]: What happened if we do not have this?

Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/package.scala]: Thanks. I'll remove this in this PR.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 2]
```diff
@@ -1148,3 +1149,128 @@ private[client] class Shim_v2_1 extends Shim_v2_0 {
...
+
+  // Statement ID
+  protected lazy val stmtIdInLoadTableOrPartition: JInteger = 0
+
+  protected lazy val listBucketingLevel: JInteger = 0
+
+  private lazy val clazzLoadFileType = getClass.getClassLoader.loadClass(
+    "org.apache.hadoop.hive.ql.plan.LoadTableDesc$LoadFileType")
+
+  private lazy val loadPartitionMethod =
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: BTW, I tracked and checked all the signature changed.

Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: Thank you, @HyukjinKwon .

**Comment by dongjoon-hyun** [comment]
The failures some from `ClassNotFoundException: org.apache.logging.log4j.util.Strings`.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by wangyum** [comment]
Can we remove the old hive support? such as 0.12, 0.13 and 0.14.

**Comment by HyukjinKwon** [comment]
Probably separate ticket for Spark 3.0.0.

**Comment by dongjoon-hyun** [comment]
I'm investigating timing issue here. Spark loads Hive Metastore class lazily. Here, Spark is trying to access Hive metastore tables like `DBS` before it's created.

**Comment by gatorsmile** [comment]
@wangyum I do not think we should deprecate the support of the previous versions of Hive metastore. Many Spark users are still using them.

**Comment by tooptoop4** [comment]
@dongjoon-hyun @wangyum @gatorsmile @HyukjinKwon Anything left on this? can it be merged to master?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala [hunk 3]
```diff
@@ -99,6 +99,7 @@ private[hive] object IsolatedClientLoader extends Logging {
     case "2.1" | "2.1.0" | "2.1.1" => hive.v2_1
     case "2.2" | "2.2.0" => hive.v2_2
     case "2.3" | "2.3.0" | "2.3.1" | "2.3.2" | "2.3.3" => hive.v2_3
+    case "3.0" | "3.0.0" => hive.v3_0
```
Comment (wangyum) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: @dongjoon-hyun Please update sql-programming-guide.md:  https://github.com/apache/spark/blob/05974f9431e9718a5f331a9892b7d81aca8387a6/docs/sql-programming-guide.md#L1217

**Comment by dongjoon-hyun** [comment]
Thank you for review, @tooptoop4 and @wangyum .
I'm going to update this to the latest Hive 3.1.0.

**Comment by tooptoop4** [comment]
bump

**Comment by dongjoon-hyun** [comment]
@tooptoop4 . Since this is a new feature, it's now targeting for Apache Spark 2.5 since `branch-2.4` is already cut. In addition, this will not be allowed for backporting `branch-2.4`. Is it urgent for you, @tooptoop4 ?

**Comment by tooptoop4** [comment]
@dongjoon-hyun I was planning to do my own custom build by cherrypicking your PR if you had it available

**Comment by dongjoon-hyun** [comment]
That would be helpful for us in the end. But, sorry for now. I'm currently not planning it open soon because the next release (Apache Spark 2.5 or 3.0) will next year.

**Comment by dongjoon-hyun** [comment]
Hi, All.
https://github.com/apache/spark/pull/23694 supercedes this PR.

**Pull request closed without merging**